src/libopensc/card-cac.c

```diff
@@ -0,0 +1,1580 @@
+
+#define CAC_MAX_SIZE 4096		/* arbitrary, just needs to be 'large enough' */
+/*
+ *  CAC hardware and APDU constants
+ */
+#define CAC_MAX_CHUNK_SIZE 240
+#define CAC_INS_GET_CERTIFICATE 0x36  /* CAC1 command to read a certificate */
+#define CAC_INS_SIGN_DECRYPT    0x42  /* A crypto operation */
+#define CAC_P1_STEP    0x80
+#define CAC_P1_FINAL   0x00
+#define CAC_INS_READ_FILE       0x52  /* read a TL or V file */
+#define CAC_FILE_TAG    1
+#define CAC_FILE_VALUE  2
+/* TAGS in a TL file */
+#define CAC_TAG_CERTIFICATE           0x70
+#define CAC_TAG_CERTINFO              0x71
+#define CAC_TAG_CUID                  0xF0
+#define CAC_TAG_CC_VERSION_NUMBER     0xF1
+#define CAC_TAG_GRAMMAR_VERION_NUMBER 0xF2
+#define CAC_TAG_CARDURL               0xF3
+#define CAC_TAG_PKCS15                0xF4
+#define CAC_TAG_ACCESS_CONTROL        0xF6
+#define CAC_TAG_DATA_MODEL            0xF5
+#define CAC_TAG_CARD_APDU             0xF7
+#define CAC_TAG_REDIRECTION           0xFA
+#define CAC_TAG_CAPABILITY_TUPLES     0xFB
+#define CAC_TAG_STATUS_TUPLES         0xFC
+#define CAC_TAG_NEXT_CCC              0xFD
+#define CAC_TAG_ERROR_CODES           0xFE
+#define CAC_APP_TYPE_GENERAL    0x01
+#define CAC_APP_TYPE_SKI        0x02
+#define CAC_APP_TYPE_PKI        0x04
+
+/* hardware data structures (returned in the CCC) */
+/* part of the card_url */
+typedef struct cac_access_profile {
+	u8 GCACR_listID;
+	u8 GCACR_readTagListACRID;
+	u8 GCACR_updatevalueACRID;
+	u8 GCACR_readvalueACRID;
+	u8 GCACR_createACRID;
+	u8 GCACR_deleteACRID;
+	u8 CryptoACR_listID;
+	u8 CryptoACR_getChallengeACRID;
+	u8 CryptoACR_internalAuthenicateACRID;
+	u8 CryptoACR_pkiComputeACRID;
+	u8 CryptoACR_readTagListACRID;
+	u8 CryptoACR_updatevalueACRID;
+	u8 CryptoACR_readvalueACRID;
+	u8 CryptoACR_createACRID;
+	u8 CryptoACR_deleteACRID;
+} cac_access_profile_t;
+
+/* part of the card url */
+typedef struct cac_access_key_info {
+	u8	keyFileID[2];
+	u8	keynumber;
+} cac_access_key_info_t;
+
+typedef struct cac_card_url {
+	u8 rid[5];
+	u8 cardApplicationType;
+	u8 objectID[2];
+	u8 applicationID[2];
+	cac_access_profile_t accessProfile;
+	u8 pinID;			     /* not used for VM cards */
+	cac_access_key_info_t accessKeyInfo; /* not used for VM cards */
+	u8 keyCryptoAlgorithm;               /* not used for VM cards */
+} cac_card_url_t;
+
+typedef struct cac_cuid {
+	u8 gsc_rid[5];
+	u8 manufacturer_id;
+	u8 card_type;
+	u8 card_id;
+} cac_cuid_t;
+
+/* data structures to store meta data about CAC objects */
+typedef struct cac_object {
+	const char *name;
+	int fd;
+	sc_path_t path;
+} cac_object_t;
+
+typedef struct cac_object_list_entry cac_object_list_entry_t;
+
+struct cac_object_list_entry {
+	cac_object_list_entry_t *next;
+	cac_object_t obj;
+};
+
+typedef struct cac_object_list {
+	cac_object_list_entry_t *head;
+	cac_object_list_entry_t *tail;
+} cac_object_list_t;
```

Comment:
I rewrote the lists in f03bff0181 to use the `simclist.h`. The code is a bit simplified:
>  3 files changed, 57 insertions(+), 76 deletions(-)

Comments welcomed.